### PR TITLE
SI-8111 Repair symbol owners after abandoned named-/default-args

### PR DESCRIPTION
**Pull request by retronym**
Names/Defaults eagerly transforms an application with temporaries
to maintain evaluation order, and dutifully changes owners of
symbols along the way.

However, if this approach doesn't work out, we throw away this
and try a auto-tupling. However, we an still witness symbols
owned by the temporaries.

This commit records which symbols are owned by the context.owner
before `transformNamedApplication`, and rolls back the changes
before `tryTupleApply`.

Perhaps a better approach would be to separate the names/defaults
applicability checks from the evaluation-order-preserving transform,
and only call the latter after we have decided to go that way.

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 1]
```diff
@@ -3269,6 +3269,13 @@ trait Typers extends Modes with Adaptations with Tags {
               // calls to the default getters. Example:
               //  foo[Int](a)()  ==>  foo[Int](a)(b = foo$qual.foo$default$2[Int](a))
               checkNotMacro()
+              def ownerOf(sym: Symbol) = if (sym == null || sym == NoSymbol) NoSymbol else sym.owner
+              val symsOwnedByContextOwner = tree.collect {
+                case t @ (_: DefTree | _: Function) if ownerOf(t.symbol) == context.owner => t.symbol
+              }
+              def rollbackNamesDefaultsOwnerChanges() {
+                symsOwnedByContextOwner foreach (_.owner = context.owner)
+              }
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:3292]: Would probably be useful to have this commented

Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:3292]: Also, I'm having hard time understanding why this isn't a no-op. You take symbols whose owner is context.owner, and then change their owners to context.owner. What effect does this achieve?

**Comment by retronym** [comment]
I'll add a comment. The owners are mutated in the NamesDefaults.

**Comment by retronym** [comment]
Comment added.

Review by @xeno-by

**Comment by xeno-by** [comment]
You alluded to the fact that it might be possible to avoid reverting. Would one of the strategies of doing that be reasonably easy to implement?

**Comment by som-snytt** [comment]
If the principled approach that avoids undoing (there's a pithy paulpism to that effect) is not required immediately, I'll take a look.  There's another named/defaults corner case issue, so it would be fun to muck around and also redeem my regression.

**Comment by retronym** [comment]
> You alluded to the fact that it might be possible to avoid reverting. Would one of the strategies of doing that be reasonably easy to implement?

Approach 1:

We could deeply duplicate trees, cloning local symbols, before the attempt at the names/defaults transform. We would throwaway that tree (and the orphaned symbols it contains) if we fall back to tupling. Looks like we have some infrastructure in place for this in `BodyDuplicator` used by specialization.

This might have a negative effect on performance.

Approach 2:

Untangle the logic to determine if a names/defaults application is applicable from the tree transformations. I'm having lunch with @lrytz today, so I'll ask for some insights about this approach.

Short answer: i don't think either approach is easy to implement.

**Comment by xeno-by** [comment]
If it's too complicated for now, let's put a TODO in the code and outline possible solutions. From what I can see, this is a critical bug, so we should probably fix it by all means necessary for now, right?

**Comment by retronym** [comment]
I marked the bug as critical on the grounds of the spectacular failure mode. But I do think the bandaid solution is a good starting point.

I've added the TODO comment.

**Comment by xeno-by** [comment]
LGTM

**Comment by xeno-by** [comment]
I think we could merge this one, right?

**Pull request merged**